### ISSUE TITLE
64 bit Windows All-In-One Binary?

### ISSUE THREAD

**Issue by RussellHaley** [luarocks/luarocks#1124]
Hi, 

Is there any plans to release a 64 bit Windows exe? I'm building a new WinLua installer that includes LuaRocks and I was hoping to produce a 64 bit variant. I speculate that LuaRocks probably doesn't care, but it would make it much much easier to manage 32 and 64 bit side-by-side installs.

I've modified the LuaRocks installer.bat to run as a configuration file generator script here: https://github.com/RussellHaley/WinLua-Mk3/tree/master/Sources/winlua-luarocks

I removed the luarocks installation and added the LUA_LIBDIR and LUA_INCDIR to the output file, as well as using setx to add the LUAROCKS_SYSCONFDIR environment variable.

My 32 bit LuaRocks installer works very well but doesn't yet autocreate the config file (next step). I'd love to have a 64 bit LuaRocks too!

Thanks for the awesome tool.
Cheers
Russ

**Comment by hishamhm** [comment]
Hi @RussellHaley,

I just took the plunge and setup a 64-bit cross-compiler environment here for Windows as well. Please let me know if these binaries work (these are builds off the latest `master`), and if they do, I'll add 64-bit Windows binaries to the next LuaRocks release:

https://hisham.hm/luarocks/luarocks-test-windows-64.zip

Thanks for the push!

**Comment by RussellHaley** [comment]
Awesome! I didn't expect this so quickly. I'll look at this after #1058 

Any chance you could share your EXE build setup? *nix is fine for me.

Rock On! (hee hee)

**Comment by hishamhm** [comment]
> Any chance you could share your EXE build setup? *nix is fine for me.

@RussellHaley it's the one that runs with `make windows-binary` (you need a mingw-w64 installation, and many of the default paths are probably tuned to my system, so they might require tweaking :) )

**Comment by RussellHaley** [comment]
I've tried `make windows-binary` on FreeBSD RELEASE-12 (using mingw32-gcc 4.8!). The build almost finished but it looks like the final build fails due to weird permissions issues?

```
russellh@jailbird:~/git/luarocks> make -j10 windows-binary-32
--- windows-binary-32 ---
gmake -f GNUmakefile windows-binary-32
gmake[1]: Entering directory '/usr/home/russellh/git/luarocks'
#gmake -f binary/Makefile.windows windows-binary MINGW_PREFIX=i686-w64-mingw32 OPENSSL_PLATFORM=mingw
gmake -f binary/Makefile.windows windows-binary MINGW_PREFIX=mingw32 OPENSSL_PLATFORM=mingw
gmake[2]: Entering directory '/usr/home/russellh/git/luarocks'
STATIC_GCC_AR=mingw32-ar \
STATIC_GCC_RANLIB=mingw32-ranlib \
STATIC_GCC_CC=mingw32-gcc \
LUAROCKS_CROSS_COMPILING=1 \
gmake binary LUA_DIR=/usr/home/russellh/git/luarocks/windows-deps-mingw32 CC=mingw32-gcc NM=mingw32-nm BINARY_PLATFORM=windows buildbinarydir=build-windows-binary-mingw32 BINARY_SYSROOT=/usr/lib/mingw-w64-sysroot/mingw32
gmake[3]: Entering directory '/usr/home/russellh/git/luarocks'
(unset ; \
"/usr/local/bin/lua53" binary/all_in_one "src/bin/luarocks" "/usr/home/russellh/git/luarocks/windows-deps-mingw32" "^src/luarocks/admin/" "/usr/local/etc/luarocks" "build-windows-binary-mingw32" "" windows mingw32-gcc mingw32-nm /usr/lib/mingw-w64-sysroot/mingw32)
(unset ; \
"/usr/local/bin/lua53" binary/all_in_one "src/bin/luarocks-admin" "/usr/home/russellh/git/luarocks/windows-deps-mingw32" "^src/luarocks/cmd/" "/usr/local/etc/luarocks" "build-windows-binary-mingw32" "" windows mingw32-gcc mingw32-nm /usr/lib/mingw-w64-sysroot/mingw32)
----------------------------------------------------------------
luasec
----------------------------------------------------------------
----------------------------------------------------------------
luasec
----------------------------------------------------------------

Error: Your user does not have write permissions in /usr/home/russellh/git/luarocks/build-windows-binary-mingw32/lua_modules/lib/luarocks/rocks-5.3
-- you may want to run as a privileged user or use your local tree with --local.
/usr/local/bin/lua53: binary/all_in_one:482: Failed building dependency: luasec
stack traceback:
        [C]: in function 'error'
        binary/all_in_one:482: in local 'main'
        binary/all_in_one:490: in main chunk
        [C]: in ?

Error: Your user does not have write permissions in /usr/home/russellh/git/luarocks/build-windows-binary-mingw32/lua_modules/lib/luarocks/rocks-5.3
-- you may want to run as a privileged user or use your local tree with --local.
/usr/local/bin/lua53: binary/all_in_one:482: Failed building dependency: luasec
stack traceback:
        [C]: in function 'error'
        binary/all_in_one:482: in local 'main'
        binary/all_in_one:490: in main chunk
        [C]: in ?
gmake[3]: *** [GNUmakefile:95: build-windows-binary-mingw32/luarocks.exe] Error 1
gmake[3]: *** Waiting for unfinished jobs....
gmake[3]: *** [GNUmakefile:99: build-windows-binary-mingw32/luarocks-admin.exe] Error 1
gmake[3]: Leaving directory '/usr/home/russellh/git/luarocks'
gmake[2]: *** [binary/Makefile.windows:19: windows-binary] Error 2
gmake[2]: Leaving directory '/usr/home/russellh/git/luarocks'
gmake[1]: *** [GNUmakefile:164: windows-binary-32] Error 2
gmake[1]: Leaving directory '/usr/home/russellh/git/luarocks'
*** [windows-binary-32] Error code 2

make: stopped in /usr/home/russellh/git/luarocks
1 error

make: stopped in /usr/home/russellh/git/luarocks
```

I manually created the directory and I still get the same error?

```
russellh@jailbird:~/git/luarocks/build-windows-binary-mingw32/lua_modules/lib/luarocks> ll
total 1
drwxr-xr-x  2 russellh  russellh  2 Jan 23 22:25 rocks-5.3/
```

So, it's likely not a permissions issue but appears like a permissions issue. I'll try building again later in debian or something.

**Comment by RussellHaley** [comment]
I've been able to build the all-in-one in a WSL Debian instance after:

-  I modified (hacked) the compiler name 
- Hacked the library/include paths.  All the mingw packages were put directly under /usr/ in this debian instance. 
- The lua-bz2 rockspec points to a private repository (hishamhm). I pointed it at the standard lua-bz2 package and it built successfully. I'm not sure if Hisham has a patch that is needed? I don't know how to exercise bz2.

I tested it with the llvm-mingw toolchain on Windows 10 (powershell) and I was able to build and use chronos. That FreeBSD error bugs me but I'm good with this.

**Comment by hishamhm** [comment]
> The lua-bz2 rockspec points to a private repository (hishamhm). I pointed it at the standard lua-bz2 package and it built successfully. I'm not sure if Hisham has a patch that is needed? I don't know how to exercise bz2.

It's not a private repo but it's a minor fork: I added some commits for compatibility across Lua versions, as I'm building the binaries using Lua 5.3: https://github.com/hishamhm/lua-bz2/commits/master

**Comment by hishamhm** [comment]
I have just released LuaRocks 3.3.0 and I shipped a 64-bit Windows binary this time!
